File: memoisation/memoisation.js
class Memoization {
    constructor () {
        this.myHashMap = new Map()
        this.simulateDelay = time => new Promise(resolve => setTimeout(resolve, time))
    }
    
    async expensiveFunction(parameter) {
        if(this.myHashMap.has(parameter)) {
            console.log('Instant response !')
            return this.myHashMap.get(parameter)
        }
    
        console.log('First time, long process...')
        await this.simulateDelay(1000)
        
        const result = `${parameter}-memoized`
    
        this.myHashMap.set(parameter, result)
    
        return result
    }

    async doStuff(){
        await this.expensiveFunction('superToto') // First time, long process...
        await this.expensiveFunction('superToto') // Instant response !
        await this.expensiveFunction('superToto') // Instant response !
        await this.expensiveFunction('superTata') // First time, long process...
        await this.expensiveFunction('superTata') // Instant response !
        await this.expensiveFunction('superToto') // Instant response !
    }
}

new Memoization().doStuff()

File: memoisation/memoisation.py
import time

class Memoization():
    def __init__(self):
        self.my_hash_map = {}

    def expensive_function (self, parameter):
        if parameter in self.my_hash_map:
            print('Instant response !')
            return self.my_hash_map[parameter]

        print('First time, long process...')
        time.sleep(1)

        result = '%s-memoized' % (parameter)

        self.my_hash_map[parameter] = result

        return result

    def do_stuff(self):
        self.expensive_function('superToto') # First time, long process...
        self.expensive_function('superToto') # Instant response !
        self.expensive_function('superToto') # Instant response !
        self.expensive_function('superTata') # First time, long process...
        self.expensive_function('superTata') # Instant response !
        self.expensive_function('superToto') # Instant response !

Memoization().do_stuff()